search_log.go

```diff
@@ -109,18 +103,65 @@ func resolveFiles(logFilePath string, beginTime, endTime int64) ([]logFile, erro
 		}
 		return nil
 	})
+
 	defer func() {
 		for _, f := range skipFiles {
 			_ = f.Close()
 		}
 	}()
+
 	// Sort by start time
 	sort.Slice(logFiles, func(i, j int) bool {
 		return logFiles[i].begin < logFiles[j].begin
 	})
 	return logFiles, err
 }
 
+func readFirstValidLog(reader *bufio.Reader, tryLines int64) (*pb.LogMessage, error) {
+	var tried int64
+	for {
+		line, err := readLine(reader)
+		if err != nil {
+			return nil, err
+		}
+		item, err := parseLogItem(line)
+		if err == nil {
+			return item, nil
+		}
+		tried++
+		if tried >= tryLines {
+			break
+		}
+	}
+	return nil, errors.New("not a valid log file")
+}
+
+func readLastValidLog(file *os.File, tryLines int) (*pb.LogMessage, error) {
+	var tried int
+	stat, _ := file.Stat()
+	endCursor := stat.Size()
+	for {
+		lines := readLastLines(file, endCursor)
+		// read out the file
+		if len(lines) == 0 {
+			break
+		}
+		endCursor -= int64(len(lines))
+		linesSlice := strings.FieldsFunc(lines, func(c rune) bool { return c == '\n' || c == '\r' })
```

Comment:
Maybe there will be problems when line ending is `\r\n`.

Reply:
Yep, I thought that, let me try to polish it.

Reply:
Polish it by `strings.Split(strings.ReplaceAll(finalStr, "\r\n", "\n"), "\n")` and put it inside the `readLastLines` method.